### ISSUE TITLE
Feature: Shorthand

### ISSUE THREAD

**Issue by STRML** [formatjs/react-intl#609]
Any opposition to supporting a shorthand syntax like other i18n libraries?

Now that v2 is out and ids are no longer hierarchical, it would be great to allow people to use message contents as ids for short, simple, common mesages.
### Syntax

`__(message: string)`
### Output

``` jsx
<FormattedMessage id={message} defaultMessage={message}>
```
### Use Case

It would be great to be able to allow things like:

``` jsx
import {__} from 'react-intl';

// ...
render() {
  return (
    <div>{__("Confirm This Order")}</div>
  )
}
```

Rather than the much more verbose:

``` jsx
render() {
  return (
    <div>
      <FormattedMessage id="common.confirmThisOrder"
        description="Order Confirmation"
        defaultMessage="Confirm This Order" />
    </div>
  )
}
```

I fully understand the value of `description`s, but there are times in many apps when it is very obvious what the context is.

`gettext` et al have used this for a long time. It's not perfect by any means but it has its place.
### Babel Plugin

I've integrated this in the babel plugin in the following commit: [babel-plugin-react-intl#6569101d](https://github.com/STRML/babel-plugin-react-intl/commit/6569101d5923338168197d2284189d6e3b552002)
### JSON Output

I'm open to comments on what the `description` should be. We should provide some context while we can. I'm thinking the file path and perhaps the containing class name, if we can get it. For now, I've just attached the file name.

``` json
[
  {
    "id": "Hello World!",
    "description": "Shorthand Message in File: test/fixtures/shorthand/actual.js",
    "defaultMessage": "Hello World!"
  },
  {
    "id": "Another message",
    "description": "Shorthand Message in File: test/fixtures/shorthand/actual.js",
    "defaultMessage": "Another message"
  }
]
```

**Comment by ericf** [comment]
There's no reason to use `__` as the export name since with ES6 `import` statements you can name the binding:

```js
import {formatMessage as __} from 'react-intl';
```

What would the return value of this function be? Are you thinking it would be something like this:

```js
export function formatMessage(message, values) {
  return <FormattedMessage id={message} defaultMessage={message} values={values}/>;
};
```

**Comment by STRML** [comment]
I think `formatMessage` would be the wrong name, given the confusion with `intlShape.formatMessage`. I chose `__` due to its very common use in other libs, but I'm open to whatever.

The return value would be exactly as you wrote.

**Comment by ericf** [comment]
I'm not on board with exporting `__` from the package. But sure, I'm open to something other than `formatMessage` which invokes the idea that you're defining and formatting a message. Maybe just `message`?

Then you could do:
```js
import {message as __} from 'react-intl';
```
Or:
```js
import {message as t} from 'react-intl';
```
…because `t` is also popular.

**Comment by STRML** [comment]
I'm partial to the gettext shorthand (which is actually `_`, not `__`, my mistake) as it's already well-known in this space. But it doesn't necessarily matter - I'm going to update the babel branch to allow configuring the names of both the long function (`defineMessages`), the shorthand function (what we're deciding), and the names of the JSX components. If a user wants to define a `global._` or inject via WebPack etc., that'll work.

How about `gettext` if we're going verbose? Too confusing?

**Comment by ericf** [comment]
> I'm going to update the babel branch to allow configuring the names of both the long function (defineMessages), the shorthand function (what we're deciding), and the names of the JSX components.

There's no need to do this. The actual exported names will be set, and Babel will automatically handle the `as` binding names.

> How about gettext if we're going verbose? Too confusing?

I don't understand why not `message`, it's clear and the user can name the binding whatever they want and the Babel plugin will work automatically.

**Comment by STRML** [comment]
> I don't understand why not `message`, it's clear and the user can name the binding whatever they want and the Babel plugin will work automatically.

I figure it should be more of a verb than a noun, but `message` is growing on me. Since it can be renamed and the babel plugin will pick up on that - :+1: from me.

**Comment by ericf** [comment]
Re: verb vs noun, yeah that's why I was thinking `formatMessage`, but I agree it could be pretty confusing since that's used else where. Since it's basically a hybrid of `defineMessage` + `formatMessage`, I think `message` can work.

Related to this, one idea I had long ago was to use ES6 tagged template literals to define and format the message. Maybe that's where this leads?

```js
<p>message`Hello, ${this.props.name}`</p>
```

…or if you renamed the import binding to `_`:
```js
<p>_`Hello, ${this.props.name}`</p>
```

Also, this relates to another discussion: https://github.com/yahoo/react-intl/issues/455#issuecomment-218282929

**Comment by STRML** [comment]
Re: interpolating the variables, that's interesting. I didn't see the value at first, but it makes sense when substituting out the message for a different locale.

We could even guess at the variable name in the babel plugin, but given this could be surprising on a refactor, it could be a different export:

```jsx
<p>{_`Hello, {${this.props.name}}`}</p>

// -->
<FormattedMessage
  id='Hello, {name}'
  defaultMessage='Hello, {name}'
  values={{
    name: 'Eric'
  }}
/>
```

**Comment by STRML** [comment]
Additionally, given the duplication in the output (and the needless increase in bundle & memory size), how do you feel about a `<FormattedSimpleMessage id="Hello, world!" />`. It would accept values like anything else, but just doesn't require the allocation of the `defaultMessage`.

Not sure if it would be worth the effort to save a tiny amount of memory.

**Comment by ericf** [comment]
There are still a few related features to this we should have ironed out and a story for:

1. Better rich-text message formatting (https://github.com/yahoo/react-intl/issues/455#issuecomment-218282929)

1. Support for shorthand formatting to a string value for use within HTML element attributes like `placeholder` which need a string value.

I'm fine with 1) being handled later. For 2), we'll need to check on the current state of React supporting a generic `<Text>` element or being able to return a string from `render()` so that the following would be supported:
```js
<input
  type='text'
  placeholder={
    <FormattedMessage
      id='foo'
      defaultMessage='Foo'
    />
  }
/>
```

Until the above can be supported by React accepting a string returned from `render()`, we'd need some story around using `message/_` shorthand for messages which we want rendered to string which is used for something like an HTML attribute value.

One way to meet 2) is that `message/_` is really shorthand for defining a message and it returns a [Message Descriptor](https://github.com/yahoo/react-intl/wiki/Components#message-descriptor) + `values` object. What would you think if the shorthand API was this:

```js
import {message as _} from 'react-intl';
// ...
<FormattedMessage {..._`Hello, {${this.props.name}}`}/>
// Using `injectIntl()`...
intl.formatMessage(_`Hello, {${this.props.name}}`); // returns a string
```

**Comment by STRML** [comment]
Why not `_` and `_s`, or `message` and `stringMessage`?

We could then pass the descriptor & values to `intl.formatMessage()` without changing its signature.

**Comment by ericf** [comment]
> Why not _ and _s, or message and stringMessage?

Because what would `stringMessage` return? It would have to be run within the intl context. So when `stringMessage` is exported from the package there isn't much it can do besides output a description of the message to be formatted.

**Comment by STRML** [comment]
Right - see my edit

**Comment by ericf** [comment]
```js
_s(intl)`Hello, {${this.props.name}}`
```

^ this will be some extra onus on the Babel plugin, and have edge cases that won't be well covered, e.g.:
```js
import React, {Component} from 'react';
import {injectIntl, messageString} from 'react-intl';

const FooIntput = ({intl}) => {
    const _s = messageString(intl);

    return (
        <input
            type='text'
            placeholder={_s`Foo bar`}
            aria-label={_s`Something helpful`}
        />
    );
};

export default injectIntl(FooIntput);
```

One thing I did to minimize this is keep the `import` right next to the message definition being extracted.

**Comment by ericf** [comment]
@STRML here's an example of what I'm thinking: https://github.com/yahoo/react-intl/pull/612

**Comment by STRML** [comment]
Should be able to make it work in the Babel plugin. It's worth it to me to keep the nice syntax vs having to use ```intl.formatMessage(_`foo`)``` and ```<FormattedMessage {..._`foo`} />```.

**Comment by ericf** [comment]
I'm worried about it being harder for people to understand though. It's uncommon for a lib to have a function that produces a React Element that can't be used with JSX syntax.

**Comment by STRML** [comment]
I agree with that sentiment. Naming can help. How about `makeFormattedMessage`?

I was also thinking that `messageString` shouldn't be an export, but a property on the `intl` object. `formatTemplateString` could be the name.

Critically, the whole point of this feature is ease-of-use - less typing means less pain when trying to migrate an app. If at all possible, I want to get the usage to be as tiny as possible. There's a lot of value to that as a "power user" option (as in, at the very end of the docs, not used in most examples), in which case it should be as minimal as possible.


```jsx
// In utils.js
export function tpl(strings, ...values) {
  const defaultMessage = strings.reduce((message, part, i) => {
    return message + part + (values[i] ? `${i}` : '');
  }, '');
  
  return {
    id: defaultMessage,
    defaultMessage,
    values,
  };
}

// In ./components/simple-message 
export default function makeFormattedMessage(strings) {
  invariant(
    Array.isArray(strings) && typeof strings[0] === 'string',
    `[React Intl] \`makeFormattedMessage\` is an ES6 tagged template function.`
  );
  // Prevents array allocation
  const {id, defaultMessage, values} = tpl.apply(this, arguments);
  return <FormattedMessage id={id} defaultMessage={defaultMessage} values={values} />
}

// In format.js, injected into intl context as a bound format function
export function messageString(config, state, ...args) {
  const descriptor = tpl(...args));
  return formatMessage(config, state, descriptor, descriptor.values);
}

// ----

// Usage
import {makeFormattedMessage as _} from 'react-intl';

const FooInput = ({intl}) => {
  const {formatTemplateString as _s} = intl;

  return (
    <span>
      <h1>_('Input Demo')</h1>
      <input type='text'
        placeholder={_s`Foo bar`}
        aria-label={_s`Something helpful`} />
    </span>
  );
};
```

**Comment by ericf** [comment]
> but a property on the intl object. formatTemplateString could be the name.

The the problem with this is making the Babel plugin correct extract the messages. Now instead of just checking if a binding references an `import` from React Intl, now you have to things through all of the scopes and people might be writing all different sorts of code or passing `intl` to other functions, etc. — essentially you'd need to implement Flow :)

> If at all possible, I want to get the usage to be as tiny as possible.

I don't agree with this goal. I think it should be as clear as possible. And I think for the `id` === `defaultMessage` case, this can be made simpler building on the existing primitives of in React Intl.

I updated the PR description with an example: https://github.com/yahoo/react-intl/pull/612

What if your code had something like this?:

```js
import {message} from 'react-intl';
import {makeFormattedMessage as _} from '../your/app/utils';
// ...
<p>_(message`Hello, {${this.props.name}}`)</p>
```

Where your app you can define `makeFormattedMessage` as simply:
```js
export const makeFormattedMessage = (descriptor) => <FormattedMessage {...descriptor}/>;
```

**Comment by STRML** [comment]
I see your point.

As a convenience, it would be very nice to have this ultra-shorthand in the lib so it can be directly translated and captured by the babel plugin without friction. If it's outside the lib, the usage would have to be ```_(message`...` ``` rather than ```_`...` ```. Would you amenable to putting `makeFormattedMessage` on `react-intl`'s exports (so it can be aliased to `_`), but also exporting `message` (or `messageTag`)?

That way you have the option to use ```_`Hello, ${this.props.name}` ```, but official examples can use ```<FormattedMessage {message`Hello, ${this.props.name}`} /> ```.

Re: the babel plugin, I'm thinking about the `{}` - I haven't looked deeply into it, but it may be able to be omitted in the general case. It would be nice to allow ````Hello, ${this.props.name}` ``` to produce `"Hello, {0}"`. 

The brackets could be kept for something interesting like ````Hello, {name ${this.props.name}]` ``` to produce `"Hello, {name}"`, and plurals. Otherwise the babel plugin could guess at it with the local var name, but it could get a little weird.

**Comment by ericf** [comment]
I really think having these additional/optional ways to accomplish the same thing would be very confusing to users.

> As a convenience, it would be very nice to have makeFormattedMessage in the lib so it can be directly translated and captured by the babel plugin without friction.

The way you create a formatted message element is `<FormattedMessage>` (which you can alias to whatever import binding you want. I feel that introducing _another_ way would be confusing.

> I'm thinking about the {} - I haven't looked deeply into it, but it may be able to be omitted in the general case. It would be nice to allow `Hello, ${this.props.name}` to produce "Hello, {0}".

I want to stick with the ICU Message syntax and I think omitting part of the syntax, sometimes, would be more harmful when someone has non-simple arguments in their message. It also becomes cumbersome in the Babel plugin to correctly detect and add missing `{}` because of whitespace, etc.

> to produce "Hello, {name}". Otherwise the babel plugin could guess at it with the local var name, but it could get a little weird.

Indexes are better here because trying to guess names based on the code can lead to too much churn. If you refactor the code, the message might be erroneously marked for re-translation — which would cost money. In fact, the whole reason why Message Descriptor has a specific `id` field is to allow for stable message ids for localization system's translation memory.

At one point I experimented with the Babel plugin generating ids as hashes, but backed away from it because it would lose the context for the translator and had the potential to lead to translation churn.

I implemented the change to Babel plugin to support the `message` template tag as a short-hand for defining a Message Descriptor: https://github.com/yahoo/babel-plugin-react-intl/pull/70

**Comment by raarts** [comment]
Disclaimer: I don't know the product, and I'm by no means an expert in i18n, but I've done my share in implementing i18n. This is just my $0.02, I'm a passer-by looking to use react-intl. Also I may be misunderstanding some of the text above.

Reading the above, I like @STRML's `_``Hello, ${this.props.name}`` a lot. I also am of the opinion that the < FormattedMessage .....> approach is too verbose for general use, visually too cluttered, and a simple shorthand is a big win.

ids can easily be auto-generated/derived from the text itself (gettext uses just the string itself as id), and a way should be found to override/supply the id where appropriate, this only happens occasionally, and would immediately make clear that there's some translation exception. 

All this to make translations as easy as possible, it's already so much work.

Thanks for listening.

**Comment by STRML** [comment]
Thanks @raarts - we've just about finished the migration in our application and have found that the `<FormattedMessage>` syntax is *very* verbose and slowed the migration significantly.

I agree that clarity is good, but if my options are increased clarity for newcomers or increased power & concision for power users, I would choose concision 100% of the time.

Even if this feature needs to be mentioned in passing at the end of the documentation as an alternative shorthand, so be it.

The use of `_` seems very simple to me and anyone who has used other i18n libraries will understand it. I understand the drive to make it more understandable, but above all I think most people would really just like to `import {_} from 'react-intl'` and have that as their primary use. As in, the point of this feature is *shorthand*. `<FormattedMessage>` is plenty clear.

Strings are of course a harder case. I think that it should be possible to track the return value of `_s(intl)` via Babel but I have not yet attempted it. It is certainly too bad that React cannot return bare strings from components.

**Comment by raarts** [comment]
Can you explain what is the problem with strings?

**Comment by STRML** [comment]
The issue is that components can extract the `intl` object from `this.context.intl` (which is needed to format strings & lookup messages). So if you have a `<FormattedMessage>` in your tree that was laid there by a component that wasn't using `intl`, that's fine.

However, if you need to add a string property to an element, like so:

```jsx
<input placeholder={<FormattedMessage id="foo" />} />
```

This won't work - React can't render directly to a string. So you have to do something like:

```jsx
<input placeholder={this.context.intl.formatMessage({id: "foo"})} />
```

which adds some significant complexity to this shorthand as it requires passing the local `intl` into it.

I would be fine with tackling that issue later and basically saying "not supported yet, just use `intl.formatMessage`. Just having `__` (like gettext) or `t` (like i18next) shorthand would be a huge win.

**Comment by raarts** [comment]
I see. And I agree.

**Comment by ericf** [comment]
> I would be fine with tackling that issue later and basically saying "not supported yet, just use intl.formatMessage. Just having __ (like gettext) or t (like i18next) shorthand would be a huge win.

@STRML @raarts it would be great to get your feedback on my variation of this proposal, which I've already implemented in #612. The difference with mine is a balance of clarity, consistency, and conciseness by having a short-hand way of defining a Message Descriptor. This way it's consistent between formatting to a React Element or string, and builds on the primitives that already exist.

**Comment by STRML** [comment]
Understood, the issue is that ```<FormattedMessage {...message`Hello, {${name}}`} />``` is still has a long way to go to be considered shorthand.

It should be:

* Usable via a single function
* As simple as possible

```<h3>{message`Hello, {${name}}`}</h3>``` is very simple and understandable. I would much rather simply ignore the string case and use `intl.formatMessage` rather than hamstring the most common use.

**Comment by ericf** [comment]
> I would much rather simply ignore the string case and use intl.formatMessage rather than hamstring the most common use.

The problem is, the string use-case is extremely common, especially if you're building an accessible application.

**Comment by ericf** [comment]
```js
<FormattedMessage {...message`Hello, {${name}}`} />
```
Remember, this can be much shorter in your app:
```js
<M {..._`Hello, {${name}}`} />
```

**Comment by STRML** [comment]
Hmm. That does make sense when displayed that way. I wish we could get it down further but context makes that very difficult. This is already a very large improvement.

:+1: Thanks for taking the time to discuss this. Let's go with it.

**Comment by raarts** [comment]
I'd really appreciate adding these shorthand types to the examples section. But it LGTM.

**Comment by oyvinmar** [comment]
I don't like the span wrapping which is needed when using the FormattedMessage component and found it a bit verbose to use injectIntl when 99% of my use cases only involve intl.formatMessage.

Ended up with creating my own HOC. It's included bellow for anyone interested.

HOC:
```javascript
export const injectT = (WrappedComponent) => {
  function getDisplayName(component) {
    return component.displayName || component.name || 'Component';
  }

  const InjectT = (props, context) => <WrappedComponent {...props} t={id => context.intl.formatMessage({ id })} />;

  InjectT.contextTypes = {
    intl: intlShape,
  };

  InjectT.displayName = `InjectT(${getDisplayName(WrappedComponent)})`;

  return InjectT;
};
```

Usage:
```javascript

import { injectT } from '...';

const WelcomePage = ({ t }) => <h1>{t('WelcomePage.helloworld')}</h1>;

export default injectT(WelcomePage);
```

**Comment by esquevin** [comment]
First, thanks for addressing my biggest concerns with react-intl.
Now, where are we on this matter ?
https://github.com/yahoo/react-intl/pull/612 has been there for a couple of month and seems to greatly reduce the complexity of using react-intl. What is currently preventing it from being merged ? When could we expect to see it merge ?

TL DR: Should I fork or should I wait ?

**Comment by ericf** [comment]
@esquevin I've been working on taking this even further to where I really want to end up with React Intl. It would be great to get your feedback on this: https://github.com/yahoo/react-intl/issues/513#issuecomment-252114827

I think it would be important to understand the overlap here, and continue to flesh out the proposal to handle rich-text messages to 1. determine if it's possible to do well, and 2. whether if make this shorthand feature pointless.

**Comment by cfey** [comment]
@ericf is there currently any way to use react-intl with the ID as optional parameter?
I got a project which has currently around 300 strings and it creates an overhead for the developers to provide an ID for every string, especially when the same string is used multiple times ("Save" as a button in a form for example)

**Comment by longlho** [comment]
You can use babel-plugin-react-intl `overrideIdFn` feature to inject ID into every string

**Comment by DragonRaider5** [comment]
Funny how a short hand like the one discussed here got rejected by two members of the community now that it's possible and even got implemented :rofl:

**Comment by jessepinho** [comment]
Picking this back up — has this feature been implemented? If it has, I can't find any reference to it in the typings or docs!
